### DISCUSSION TITLE
Miss-use of $value/$label in your Filter class

### DISCUSSION THREAD

**Discussion by sfroemkenjw** [laravel/nova-issues#6022] [Q&A] [answered]
- Laravel Version: 10.20.0
- Nova Version: 4.27.12
- PHP Version: 8.2.9
- Database Driver & Version: MariaDB 10.4
- Operating System and Version: ddev version v1.22.2 on Mac with Docker Desktop for Mac.
- Browser type and version: Firefox 119.0

### Description:

In list view of users we have a dropdown to filter users by their roles. This selectbox was build with equal `value` and `label` in HTML:

- `<option value="BusinessPartner DE">BusinessPartner DE</option>`
- `<option value="BusinessPartner EN">BusinessPartner EN</option>`
- `<option value="BusinessPartner FR">BusinessPartner FR</option>`
...

instead of:

- `<option value="2">BusinessPartner DE</option>`
- `<option value="5">BusinessPartner EN</option>`
- `<option value="12">BusinessPartner FR</option>`
...

The `options` method of our `UserGroupFilter`

```
/**
 * Get the filter's available options.
 */
public function options(NovaRequest $request): array
{
    return Role::query()->orderBy('name')->pluck('name', 'id')->toArray();
}
```

returns the correct array:

```
array(
    2 => "BusinessPartner DE",
    5 => "BusinessPartner EN",
    12 => "BusinessPartner FR"
)
```

I have xdebuged the return value. The array keys and values are correct until now. In your `Filter` class I found the calling part:

```
'options' => collect($this->options(app(NovaRequest::class)))->map(function ($value, $label) {
    if (is_array($value)) {
        return array_merge(['label' => $label], $value);
    } elseif (is_string($label)) {
        return ['label' => $label, 'value' => $value];
    }

    return ['label' => $value, 'value' => $value];
})->values()->all(),
```

You use my `array` and put it into your `Collection` class. Here, the data is still correct. But now, you apply a callable to each array entry. This will be done in Laravels `Arr` class:

```
$items = array_map($callback, $array, $keys);
```

So, first parameter for your `callable` is the `label` (`BusinessPartner X`) and second parameter is the `value` (`2`, `5`, `12`, ...)

But in your callable you use `$value` as `label` and `$label` as `value`. So, `$label` is f.e. `2 (int)` which is NOT `string` which results in processing the HTML `option` with same value for `value` and `label`:

`return ['label' => $value, 'value' => $value];`

Would be cool, if you can change `$value` with `$label` and adopt the code accordingly.

Nice greetings

Stefan

**Comment by crynobone** [comment, marked as answer]
You can change the format to the following:

```php
return Role::query()->orderBy('name')->pluck('id', 'name')->toArray();
```

GIven the following example:

```php
[
    "BusinessPartner DE" => 2,
    "BusinessPartner EN" => 5,
    "BusinessPartner FR" => 12,
    "Partner France" => 12,
]
```

vs

```php
[
    2 => "BusinessPartner DE",
    5 => "BusinessPartner EN",
    12 => "BusinessPartner FR",
    12 => "Partner France"
]
```

With labels as key, you can have multiple redundant values with different labels. Having value as the key eliminates this possibility.

**Reply by sfroemkenjw** [reply]
Hello @crynobone 

as the array comes from a database I hope that `value` (`ID`) will never be redundant ;-)
The official description of second parameter of `pluck` is `string $key` which sounds like an unique identifier for my array entries.
Currently, I interprete your solution as a temporary quick-fix.

Stefan

**Reply by crynobone** [reply]
I'm explaining to you why it is this way. No plan to change it on minor releases.
